pkg/querier/worker/scheduler_processor.go

```diff
@@ -98,17 +98,12 @@ func (sp *schedulerProcessor) notifyShutdown(ctx context.Context, conn *grpc.Cli
 	}
 }
 
-func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Context, conn *grpc.ClientConn, address string) {
+func (sp *schedulerProcessor) processQueriesOnSingleStream(ctx context.Context, conn *grpc.ClientConn, address string) {
 	schedulerClient := sp.schedulerClientFactory(conn)
 
-	// Run the querier loop (and so all the queries) in a dedicated context that we call the "execution context".
-	// The execution context is cancelled once the workerCtx is cancelled AND there's no inflight query executing.
-	execCtx, execCancel, inflightQuery := newExecutionContext(workerCtx, sp.log)
-	defer execCancel()
-
-	backoff := backoff.New(execCtx, processorBackoffConfig)
+	backoff := backoff.New(ctx, processorBackoffConfig)
 	for backoff.Ongoing() {
-		c, err := schedulerClient.QuerierLoop(execCtx)
```

Comment:
I've introduced the execution context. The reason at that time was that we had to pass a different context (the execution one) to `schedulerClient.QuerierLoop()` too, otherwise the returned gRPC stream client will get canceled when the context passed to `processQueriesOnSingleStream()` is cancelled.

The context passed to `processQueriesOnSingleStream()` is cancelled when the querier shutdown, but if we cancel it right away then the subsequent calls to the gRPC stream client will fail, while they shouldn't because we're still processing the in-flight queries and we need to report back the result to the query-scheduler.

Am I missing anything?

Reply:
Ah, I hadn't considered that scenario.

These are the two groups of scenarios I understand we need to cater for - have I missed any others?

Query-frontend cancels a query: 
* This is currently communicated to queriers by the query-scheduler sending an error in response to a querier's `Recv()` call and closing the stream. 
* The inflight query (if any) should stop, and the result or cancellation does not need to be reported to the frontend, but it is safe to report it nonetheless.
* Notifying the scheduler that the query was completed is not required.
* The querier should reconnect to the scheduler to continue processing further queries. 

Querier discovers that a query-scheduler is shutting down, or querier is shutting down: 
* The pending `Recv()` call to the scheduler should be stopped.
* The inflight query (if any) should run to completion as per normal and report the result to the frontend. 
* Notifying the scheduler that the query was completed is required, otherwise the scheduler might tell the frontend the query failed.

As you point out, the behaviour for this second group isn't quite right with this PR: it won't report back to the scheduler in this case. 

I can see two possible solutions:

* We can ignore the race that this PR tries to fix, and rely on #6703 to report the cancellation back to the frontend, which will then retry the query.
* We change the behaviour for the second group so that the `Recv()` call is stopped not by cancelling the context on the querier side, but by the scheduler. This could be done by the querier notifying the scheduler that it wants to terminate the stream as a separate RPC, independent of the `QuerierLoop` stream, and the scheduler sending an acknowledgement to the pending `Recv()` call. This second option would remove the unnecessary retry, but introduce a bit of complexity.

What do you think? Any other ideas?

Reply:
I've got a rough sketch of the second idea in ba628ca (the tests are failing, will fix them if we decide to pursue this). I don't love it, but I don't hate it either.

On the one hand, it fixes the race here, and simplifies some of the logic. It also seems straightforward to do in a backwards compatible way, so we can safely handle combinations of old and new queriers and query-schedulers.

On the other hand, it makes other things more complex, and likely introduces new and exciting failure modes I haven't even considered.

Reply:
> These are the two groups of scenarios I understand we need to cater for - have I missed any others?

I think what you describe covers it.

> Query-frontend cancels a query:
> This is currently communicated to queriers by the query-scheduler sending an error in response to a querier's Recv() call and closing the stream.

Doesn't the query-scheduler simply cancels the querier's gRPC stream context to signal to cancel the in-flight query when it gets cancelled by the query-frontend?

> We can ignore the race that this PR tries to fix, and rely on https://github.com/grafana/mimir/pull/6703 to report the cancellation back to the frontend, which will then retry the query.

Given the complexity of all this logic and the rare case of the race, I would be tempted to not touch such a complicated logic and just rely on this ☝️

Reply:
> We change the behaviour for the second group so that the Recv() call is stopped not by cancelling the context on the querier side, but by the scheduler. This could be done by the querier notifying the scheduler that it wants to terminate the stream as a separate RPC

Looks an over complication of an already-complicated and fragile workflow.

Reply:
> > Query-frontend cancels a query:
> > This is currently communicated to queriers by the query-scheduler sending an error in response to a querier's Recv() call and closing the stream.
> 
> Doesn't the query-scheduler simply cancels the querier's gRPC stream context to signal to cancel the in-flight query when it gets cancelled by the query-frontend?

Both: it returns an error [here](https://github.com/grafana/mimir/blob/18a4e2bcaa044c286ef4c568553f3c3ff5591163/pkg/scheduler/scheduler.go#L481) which is then communicated back to the client and would be received by the querier [here](https://github.com/grafana/mimir/blob/18a4e2bcaa044c286ef4c568553f3c3ff5591163/pkg/querier/worker/scheduler_processor.go#L147). The server sending an error closes the stream, and the gRPC client cancels the context associated with the client-side of the stream when the error is received.

Reply:
>Looks an over complication of an already-complicated and fragile workflow.

Marco and I discussed this offline - we've agreed to rely on the retry behaviour from #6703 to cover this scenario, so I'm going to close this.